aodntools/timeseries_products/gridded_timeseries.py

```diff
@@ -0,0 +1,262 @@
+    :return: list of depth bins
+    """
+
+    dmin = int(nc.DEPTH.min() / increment) * increment
+    dmax = int(nc.DEPTH.max() / increment) * increment + increment
+    dbin = list(range(dmin, dmax, increment))
+
+    return dbin
+
+def adjust_depth_bins(depth_bins, depth_min, depth_max):
+    """
+    Adjust the provided depth bins to match the min and max registered depth
+    :param depth_bins: list of target depths
+    :param depth_min: min valid depth
+    :param depth_max: max valid depth
+    :return: list of adjusted target depths
+    """
+
+    target_depth_min = bisect.bisect_left(depth_bins, depth_min)
+    target_depth_max = bisect.bisect_right(depth_bins, depth_max)
+    return depth_bins[target_depth_min:target_depth_max]
+
+
+def get_depth_mask(depth_bins, depths, max_separation):
+    """
+    return a boolean mask of depth bins where the interpolation is possible
+    based on maximum separation between the instruments
+    :param depth_bins: list of depth bins where the interpolation will happen
+    :param depths: list of depth of the instruments at the current timestamp
+    :param max_separation: maximum allowable separation between readings
+    :return: list of depth masked
+    """
+    depth_mask = [False] * len(depth_bins)
+    for depth_index in range(len(depths) - 1):
+        if depths[depth_index + 1] - depths[depth_index] < max_separation:
+            depth_min = bisect.bisect_left(depth_bins, depths[depth_index])
+            depth_max = bisect.bisect_right(depth_bins, depths[depth_index + 1])
+            depth_interval_len = depth_max - depth_min
+            depth_mask[depth_min:depth_max] = [True] * depth_interval_len
+    return depth_mask
+
+def sort_attributes(attributes):
+    """
+
+    :param attributes: dictionary of global attributes
+    :return: sorted dictionary
+    """
+    return dict(sorted(attributes.items()))
+
+def sort_depths(depths, values):
+    """
+    Sort the list of depths and values
+    """
+    index =list(range(len(depths)))
+    index.sort(key=depths.__getitem__)
+    depths[:] = [depths[i] for i in index]
+    values[:] = [values[i] for i in index]
+
+    return depths, values
+
+
+
+
+## MAIN FUNCTION
+def grid_variable(file_name, VoI, depth_bins=None, max_separation=50, depth_bins_increment=10,
+                  base_path='./'):
+    """
+    Grid VoI into depth_bins.
+    :param nc: hourly aggregated dataset with VoI, DEPTH and TIME only
+    :param VoI: variable of interest (TEMP or PSAL)
+    :param depth_bins: list of depth where to interpolate. if null list is provided it will be calculated from the data
+    :param max_separation: max separation allowed for instruments
+    :param depth_bins_increment: in case no depth bins provided this is the increment for the calculated bins
+    :param base_path: path where the result file will be written
+    :return: interpolated dataset
+    """
+
+    with xr.open_dataset(file_name) as nc:
+        site_code = nc.site_code
```

Comment:
If you must, all those dictionary access could be checked before hand and hence, bail out before computing stuff - for example:

```
required_keys = ['site_code','LONGITUDE','LATITUDE','etc']
kerror = [f"key={key} not found\n") for key in nc.keys() if key not in required_keys]
if kerror:
raise KeyError(''.join(kerror))
end
```

You could even wrap that in a function to be called before everything else so you know beforehand which files are incomplete.

Reply:
not needed as we're using the hourly aggregated times series product as input, which is completely under our control.